src/comm.ts

```diff
@@ -0,0 +1,379 @@
+   * Unregister a comm instance.
+   */
+  private _unregisterComm(commId: string) {
+    this._comms.delete(commId);
+  }
+
+  private _comms = new Map<string, Promise<IComm>>();
+  private _kernel: IKernel = null;
+  private _targets = new Map<string, (comm: IComm, data: any) => any>();
+}
+
+
+/**
+ * Comm channel handler.
+ */
+export
+class Comm extends DisposableDelegate implements IComm {
+
+  /**
+   * Construct a new comm channel.
+   */
+  constructor(targetName: string, commId: string, kernel: IKernel, callback: () => void) {
+    super(callback);
+    this._target = targetName;
+    this._id = commId;  
+    this._kernel = kernel;
+  }
+
+  /**
+   * Get the uuid for the comm channel.
+   *
+   * Read-only
+   */
+  get commId(): string {
+    return this._id;
+  }
+
+  /** 
+   * Get the target name for the comm channel.
+   *
+   * Read-only
+   */
+  get targetName(): string {
+    return this._target;
+  }
+
+  /** 
+   * Get the onClose handler.
+   */
+  get onClose(): (data?: any) => void {
+    return this._onClose;
+  }
+
+  /**
+   * Set the onClose handler.
+   */
+  set onClose(cb: (data?: any) => void) {
+    this._onClose = cb;
+  }
+
+  /**
+   * Get the onMsg handler.
+   */
+  get onMsg(): (data: any) => void {
+    return this._onMsg;
+  }
+
+  /**
+   * Set the onMsg handler.
+   */
+  set onMsg(cb: (data: any) => void) {
+    this._onMsg = cb;
+  }
+
+  /**
+   * Initialize a comm with optional data.
+   */
+  open(data?: any, metadata?: any): IKernelFuture {
+    var contents = {
+      comm_id: this._id,
+      target_name: this._target,
+      data: data || {}
+    }
+    return sendCommMessage(this._kernel, 'comm_open', contents, metadata);
+  }
+
+  /**
+   * Send a comm message to the kernel.
+   */
+  send(data: any, metadata={}, buffers: (ArrayBuffer | ArrayBufferView)[]=[]): IKernelFuture {
+    if (this.isDisposed) {
+      throw Error('Comm is closed');
+    }
+    var contents = { comm_id: this._id, data: data };
+    return sendCommMessage(this._kernel, 'comm_msg', contents, metadata, buffers);
+  }
+
+  /**
+   * Close the comm.
+   */
+  close(data?: any, metadata={}): IKernelFuture {
+    if (this.isDisposed) {
+      return;
+    }
+    var onClose = this._onClose;
+    if (onClose) onClose(data);
+    var contents = { comm_id: this._id, data: data || {} };
+    var future = sendCommMessage(this._kernel, 'comm_close', contents, metadata);
+    this._onClose = null;
+    this._onMsg = null;
+    this.dispose();
+    return future;
+  }
+
+  private _target = '';
+  private _id = '';
+  private _kernel: IKernel = null;
+  private _onClose: (data?: any) => void = null;
+  private _onMsg: (data: any) => void = null;
+}
+
+
+/**
+ * Load a target from a module using require.js, if a module 
+ * is specified, otherwise try to load a target from the given registry.
+ */
+function loadTarget(targetName: string, moduleName: string, registry: Map<string, (comm: IComm, data: any) => any>): Promise<any> {
```

Comment:
Nevermind, it was supposed to be a callback: `Promise<(comm: IComm, data: any) => any>`